<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Get candidates non-promoter cis regulation — cis_em_genomic • methylayer</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Get candidates non-promoter cis regulation — cis_em_genomic" />
<meta property="og:description" content="Get candidates non-promoter cis regulation" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">methylayer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get candidates non-promoter cis regulation</h1>
    
    <div class="hidden name"><code>cis_em_genomic.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Get candidates non-promoter cis regulation</p>
    </div>

    <pre class="usage"><span class='fu'>cis_em_genomic</span><span class='op'>(</span>
  <span class='va'>meth_mat</span>,
  <span class='va'>expr_mat</span>,
  <span class='va'>gene_tss</span>,
  min_samples <span class='op'>=</span> <span class='fl'>100</span>,
  max_k <span class='op'>=</span> <span class='fl'>50</span>,
  max_dist <span class='op'>=</span> <span class='fl'>500000</span>,
  min_dist <span class='op'>=</span> <span class='fl'>200</span>,
  spearman <span class='op'>=</span> <span class='cn'>FALSE</span>,
  parallel <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>meth_mat</th>
      <td><p>Matrix with methylation values. Each row is a locus and each column is a sample. Rownames should contain "chrom", "start" and "end" separated by "_"</p></td>
    </tr>
    <tr>
      <th>expr_mat</th>
      <td><p>Matrix with expression values. Each row is a gene and each column is a sample.</p></td>
    </tr>
    <tr>
      <th>gene_tss</th>
      <td><p>intervals set with additional "name" column with transcription start site (TSS) for each gene. Used for allocating a genomic coordinate for each gene. In case of duplicate names or coordinates the first one in the matrix will be used.</p></td>
    </tr>
    <tr>
      <th>min_samples</th>
      <td><p>minimal number of samples per gene (both expression and methylation). Default is 100.</p></td>
    </tr>
    <tr>
      <th>max_k</th>
      <td><p>maximal number of genes to report per locus. Default is 50. Note that for efficiency (memory and computation) it is best to keep this number small.</p></td>
    </tr>
    <tr>
      <th>max_dist</th>
      <td><p>maximal distance from TSS to consider "cis" (used for FDR computations)</p></td>
    </tr>
    <tr>
      <th>min_dist</th>
      <td><p>minimal distance from TSS (in order to exclude promoter regulation). Used for FDR computation.</p></td>
    </tr>
    <tr>
      <th>spearman</th>
      <td><p>use spearman correlation (if FALSE - use pearson)</p></td>
    </tr>
    <tr>
      <th>parallel</th>
      <td><p>use mclapply for computation of ranks. Note that for large matrices forking might fail due to lack of memory and therefore parallel should be set to FALSE.</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aviezer Lifshitz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


